cardano-crypto-class/src/Cardano/Crypto/DSIGN/SchnorrSecp256k1.hs

```diff
@@ -0,0 +1,296 @@
+{-# OPTIONS_GHC -fno-full-laziness #-}
+
+module Cardano.Crypto.DSIGN.SchnorrSecp256k1 (
+  SchnorrSecp256k1DSIGN,
+  VerKeyDSIGN,
+  SignKeyDSIGN,
+  SigDSIGN
+  ) where
+
+import qualified Data.ByteString as BS
+import Data.ByteString.Unsafe (unsafePackCStringLen)
+import Data.Primitive.Ptr (copyPtr)
+import Crypto.Random (getRandomBytes)
+import Cardano.Crypto.Seed (runMonadRandomWithSeed)
+import Data.ByteString.Internal (toForeignPtr, memcmp)
+import Foreign.ForeignPtr (
+  ForeignPtr, 
+  withForeignPtr, 
+  mallocForeignPtrBytes,
+  plusForeignPtr, 
+  castForeignPtr
+  )
+import Control.Monad (when)
+import System.IO.Unsafe (unsafeDupablePerformIO, unsafePerformIO)
+import Cardano.Binary (FromCBOR (fromCBOR), ToCBOR (toCBOR, encodedSizeExpr))
+import Foreign.Ptr (Ptr, castPtr, nullPtr)
+import Foreign.C.Types (CUChar)
+import Foreign.Marshal.Alloc (allocaBytes)
+import Cardano.Crypto.Schnorr (
+  SECP256k1XOnlyPubKey, 
+  secpKeyPairCreate, 
+  SECP256k1Context, 
+  secpKeyPairXOnlyPub,
+  SECP256k1SecKey, 
+  secpSchnorrSigVerify, 
+  secpContextSignVerify,
+  SECP256k1SchnorrSig, 
+  secpSchnorrSigSignCustom, 
+  secpContextCreate
+  )
+import NoThunks.Class (NoThunks, OnlyCheckWhnfNamed (OnlyCheckWhnfNamed))
+import Cardano.Crypto.DSIGN.Class (
+  DSIGNAlgorithm (VerKeyDSIGN, 
+                  SignKeyDSIGN, 
+                  SigDSIGN,
+                  SeedSizeDSIGN, 
+                  SizeSigDSIGN, 
+                  SizeSignKeyDSIGN, 
+                  SizeVerKeyDSIGN, 
+                  algorithmNameDSIGN,
+                  deriveVerKeyDSIGN, 
+                  signDSIGN, 
+                  verifyDSIGN, 
+                  genKeyDSIGN, 
+                  rawSerialiseSigDSIGN,
+                  Signable, 
+                  rawSerialiseVerKeyDSIGN, 
+                  rawSerialiseSignKeyDSIGN, 
+                  rawDeserialiseVerKeyDSIGN,
+                  rawDeserialiseSignKeyDSIGN, 
+                  rawDeserialiseSigDSIGN), 
+  encodeVerKeyDSIGN, 
+  encodedVerKeyDSIGNSizeExpr, 
+  decodeVerKeyDSIGN, 
+  encodeSignKeyDSIGN, 
+  encodedSignKeyDESIGNSizeExpr, 
+  decodeSignKeyDSIGN, 
+  encodeSigDSIGN, 
+  encodedSigDSIGNSizeExpr, 
+  decodeSigDSIGN
+  )
+import Cardano.Crypto.Util (SignableRepresentation (getSignableRepresentation))
+
+data SchnorrSecp256k1DSIGN
+
+instance DSIGNAlgorithm SchnorrSecp256k1DSIGN where
+  type SeedSizeDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeSigDSIGN SchnorrSecp256k1DSIGN = 64
+  type SizeSignKeyDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeVerKeyDSIGN SchnorrSecp256k1DSIGN = 64
+  type Signable SchnorrSecp256k1DSIGN = SignableRepresentation
+  newtype VerKeyDSIGN SchnorrSecp256k1DSIGN = 
+    VerKeySchnorr256k1 (ForeignPtr SECP256k1XOnlyPubKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "VerKeySchnorr256k1" (ForeignPtr SECP256k1XOnlyPubKey)
+        )
+  newtype SignKeyDSIGN SchnorrSecp256k1DSIGN = 
+    SignKeySchnorr256k1 (ForeignPtr SECP256k1SecKey)
```

Comment:
Why not use `PinnedSizedBytes` for these?

Reply:
For the same reason as mentioned previously. My experience with FFI tells me that it's much better to be closer to C, and 'calling down' type level numbers to the value level is _not_ a cheap operation.

Reply:
Again, this is false: look at the Core if you don't believe me. These are statically known, so there is no overhead.